### PR TITLE
multi-image support for llama3.2

### PR DESCRIPTION
**Pull request by yma11**
FILL IN THE PR DESCRIPTION HERE

FIX #xxxx (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ https://docs.vllm.ai/en/latest/contributing/overview.html **

### CODE REVIEW

**Comment by yma11** [comment]
@kdamaszk @michalkuligowski @kzawora-intel, can you help review on this PR? Already tested on 11B and 90B.

File: vllm/worker/hpu_enc_dec_model_runner.py [hunk 1]
```diff
@@ -355,112 +355,82 @@ def _prepare_encoder_model_input_tensors(
...
+        kv_caches = [
+            torch.tensor([], dtype=torch.float32, device=self.device)
+            for _ in range(num_layers)
+        ]
+        finished_requests_ids = [seq.request_id for seq in seqs]
+        model_input = self.prepare_model_input(
+            seqs, finished_requests_ids=finished_requests_ids)
+        intermediate_tensors = None
+        self.execute_model(model_input, kv_caches, intermediate_tensors)
+        torch.cuda.synchronize()
```
Comment (kdamaszk) [inline vllm/worker/hpu_enc_dec_model_runner.py:432]: This won't work on HPU :)

Comment (yma11) [reply vllm/worker/hpu_enc_dec_model_runner.py:432]: oh, sorry for a mistake code push. let me update it.

Comment (yma11) [reply vllm/worker/hpu_enc_dec_model_runner.py]: Done.

File: vllm/worker/hpu_enc_dec_model_runner.py [hunk 2]
```diff
@@ -355,112 +355,84 @@ def _prepare_encoder_model_input_tensors(
...
-                          max_seq_len)
-
-        self.warmup_scenario(max_batch_size, max_seq_len, True, kv_caches,
-                             False)
-        return
+        # Enable top-k sampling to reflect the accurate memory usage.
+        sampling_params = SamplingParams(top_p=0.99, top_k=self.vocab_size - 1)
+        max_num_batched_tokens = self.scheduler_config.max_num_batched_tokens
+        # Workaround to avoid unexpeced OOM failure during profile run
+        max_num_seqs = int(self.scheduler_config.max_num_seqs/2)
```
Comment (kdamaszk) [inline vllm/worker/hpu_enc_dec_model_runner.py:362]: I don't like this. Max prefill batch size is limited by `self.max_num_prefill_seqs`. I believe the role of `profile_run` is to generate with maximum possible batch size. If it is too big than user should limit it. Such hacks will not prevent from actual OOM in the middle of benchmark.

Comment (yma11) [reply vllm/worker/hpu_enc_dec_model_runner.py:362]: Understand your points. This is a temporary work around. For multi-modal models, this configurations not only affect the generated prefill prompts but also the encode dummy data, which will occupy a lot of memory during process before actually computed together with prefill and cause OOM. Current encode data process has some flaws and will be our next step to see how to improve it. Also, I have tested different configurations on benchmarks and not observed actual OOM.

Comment (yma11) [reply vllm/worker/hpu_enc_dec_model_runner.py:362]: @kdamaszk QA has validated some configurations based on this PR. Can we merge it first and let me try addressing your point in my later PRs?

Comment (yma11) [reply vllm/worker/hpu_enc_dec_model_runner.py]: @kdamaszk I revert back to use previous logic but add some fixes for image tokens. Can you help merge this PR if there is no other big issue? QA need this PR merged to launch test. I can address in my later PR if there are some minor comments.

Comment (kdamaszk) [reply vllm/worker/hpu_enc_dec_model_runner.py]: It looks like there are no any accuracy and performance regression with your change, so lgtm

File: vllm/worker/hpu_enc_dec_model_runner.py [hunk 3]
```diff
@@ -471,18 +476,30 @@ def create_dummy_seq_group_metadata(self,
                                    max_mm_tokens) // self.block_size
             cross_block_table = [_PAD_BLOCK_ID] * num_cross_blocks
         prompt_token_ids = [0] * input_len
+        if is_prompt:
+            image_data = encoder_dummy_data.multi_modal_data["image"]
+            if isinstance(image_data, Image.Image):
+                image_data = [image_data]
+            assert is_list_of(image_data, Image.Image)
+            text_prompt_len = input_len - 2 - len(image_data)
+            prompt_token_ids = [128000] + [128256] * len(image_data) + [
```
Comment (kdamaszk) [inline vllm/worker/hpu_enc_dec_model_runner.py:485]: It would be good to provide some justification about "magic numbers" here.

**Comment by kdamaszk** [comment]
@yma11 I observed accuracy regression on MMMU val dataset and LLama 3.2 11B Vision Instruct. Let's sync offline

File: vllm/worker/hpu_enc_dec_model_runner.py (hunk 3 above)
Comment (yma11) [reply vllm/worker/hpu_enc_dec_model_runner.py:487]: Added.

Comment (kdamaszk) [reply vllm/worker/hpu_enc_dec_model_runner.py:487]: line too long :( requirement for max 80 is pretty low, but we need to follow this

Comment (yma11) [reply vllm/worker/hpu_enc_dec_model_runner.py:487]: Thought it's 100. Updated.

**Comment by kdamaszk** [comment]
@PatrykWo please review

**Comment by PatrykWo** [comment]
@kdamaszk I see some errors in CI. Please rebase with the base branch so you are up to date and we can confirm that new changes does not influence the PR.

**Comment by yma11** [comment]
@PatrykWo I did a rebase to latest branch but CI may still fail based on @kdamaszk 's comments. Can you help take a further look?

File: vllm/model_executor/models/mllama.py [hunk 4]
```diff
@@ -889,14 +909,21 @@ def _attention_with_mask(
...
+            return output
+        else:
+            output = F.scaled_dot_product_attention(q,
+                                                    k,
+                                                    v,
+                                                    attn_mask=attention_mask,
+                                                    is_causal=False)
+            output = output.permute(2, 0, 1, 3).reshape(
+                q_len, self.num_local_heads * self.head_dim)
+            return output
```
Comment (jkaniecki) [inline vllm/model_executor/models/mllama.py:926]: @yma11 This PR works fine but using Fused SDPA in cross attetnion results in worse accuracy. We should better keep F.scaled_dot_product_attention here

**Review by jkaniecki — CHANGES_REQUESTED** [review]
I've tested the multi image support and it seems to work properly, but whit Fused SDPA in cross attention layers accuracy is worse - please, revert changes made in SDPA computation

File: vllm/model_executor/models/mllama.py (hunk 4 above)
Comment (yma11) [reply vllm/model_executor/models/mllama.py]: Done. But note that `F.scaled_dot_product_attention` consumes more memory than Fused SDPA.

File: vllm/worker/hpu_enc_dec_model_runner.py [hunk 5]
```diff
@@ -9,6 +9,7 @@
 
 import habana_frameworks.torch as htorch
 import torch
+from PIL import Image
```
Comment (michalkuligowski) [inline vllm/worker/hpu_enc_dec_model_runner.py:12]: Please fix merge conflicts

**Review by michalkuligowski — CHANGES_REQUESTED** [review]

File: vllm/model_executor/models/mllama.py [hunk 6]
```diff
@@ -929,14 +947,33 @@ def _attention_with_mask(
...
         attention_mask = attention_mask.view(1, 1, q_len, kv_len)
-        output = F.scaled_dot_product_attention(q,
-                                                k,
-                                                v,
-                                                attn_mask=attention_mask,
-                                                is_causal=False)
-        output = output.permute(2, 0, 1, 3).reshape(
-            q_len, self.num_local_heads * self.head_dim)
-        return output
+        if current_platform.is_hpu():
```
Comment (jkaniecki) [inline vllm/model_executor/models/mllama.py:950]: No need to permute and return output in both "if" and "else" - let's permute and return output after the "if" as now we perform code duplication

File: vllm/model_executor/models/mllama.py [hunk 7]
```diff
@@ -1001,9 +1038,13 @@ def forward(
...
-        if len(hidden_states.shape) == 3:
-            full_text_row_masked_out_mask = full_text_row_masked_out_mask.view(
-                hidden_states.size(0), -1, 1)
+        assert len(residual.shape) == 3
+        if len(hidden_states.shape) == 2:
+            hidden_states = hidden_states.view(residual.size(0),
+                                               residual.size(1),
+                                               residual.size(2))
+        full_text_row_masked_out_mask = full_text_row_masked_out_mask.view(
+            hidden_states.size(0), -1, 1)
```
Comment (jkaniecki) [inline vllm/model_executor/models/mllama.py:1047]: If this is only hpu related we should put it under some "if". We cannot delete original code here if it's different from our one.

File: vllm/model_executor/models/mllama.py [hunk 8]
```diff
@@ -1320,7 +1361,12 @@ def get_cross_attention_mask(
         num_tokens_per_tile: int,
         dtype: torch.dtype,
     ) -> Tuple[torch.Tensor, torch.Tensor]:
-        token_ids = input_ids.tolist()
+        token_ids = []
```
Comment (jkaniecki) [inline vllm/model_executor/models/mllama.py:1364]: Why we need this empty list here?

File: vllm/worker/hpu_enc_dec_model_runner.py [hunk 9]
```diff
@@ -446,52 +448,50 @@ def create_dummy_seq_group_metadata(self,
...
         return SequenceGroupMetadata(
             request_id=str(group_id),
             is_prompt=is_prompt,
             seq_data={group_id: seq_data},
             sampling_params=sampling_params,
             block_tables=block_tables,
             encoder_seq_data=encoder_dummy_data.seq_data,
-            multi_modal_data=decoder_dummy_data.multi_modal_data,
+            multi_modal_data=decoder_dummy_data.multi_modal_data
+            or encoder_dummy_data.multi_modal_data,
```
Comment (jkaniecki) [inline vllm/worker/hpu_enc_dec_model_runner.py:491]: Why putting "or encoder_dummy_data.multi_modal_data" ? In our case encoder_dummy_data always has no multimodal data

File: vllm/worker/hpu_enc_dec_model_runner.py [hunk 10]
```diff
@@ -446,52 +448,50 @@ def create_dummy_seq_group_metadata(self,
...
             multi_modal_placeholders=decoder_dummy_data.
-            multi_modal_placeholders,
+            multi_modal_placeholders
```
Comment (jkaniecki) [inline vllm/worker/hpu_enc_dec_model_runner.py:493]: Why putting "or encoder_dummy_data.multi_modal_placeholders" ? In our case encoder_dummy_data always has no multimodal data

**Review by jkaniecki — CHANGES_REQUESTED** [review]

File: vllm/model_executor/models/mllama.py (hunk 7 above)
Comment (yma11) [reply vllm/model_executor/models/mllama.py:1044]: The code change here doesn't change the process logic and if the `if` condition happens in other platforms, it should do same thing.

**Review by jkaniecki — APPROVED** [review]

**Comment by yma11** [comment]
@michalkuligowski Can you help take a review on this PR?

File: vllm/model_executor/models/mllama.py [hunk 11]
```diff
@@ -1001,7 +1035,12 @@ def forward(
         # TODO: Change input_tokens tensor at the beginning of model execution
         # to 2D tensor to align with public vllm input_tokens shape. But this
         # will face the graph building failure issue, still need to investigate.
-        if len(hidden_states.shape) == 3:
+        if current_platform.is_hpu():
```
Comment (michalkuligowski) [inline vllm/model_executor/models/mllama.py:1038]: if is_hpu:

Comment (yma11) [reply vllm/model_executor/models/mllama.py]: @michalkuligowski updated. Thanks.

File: vllm/worker/hpu_enc_dec_model_runner.py [hunk 12]
```diff
@@ -358,9 +356,13 @@ def _prepare_encoder_model_input_tensors(
 
         return attn_metadata
 
+    @torch.inference_mode()
```
Comment (michalkuligowski) [inline vllm/worker/hpu_enc_dec_model_runner.py:359]: Why is inference mode set here for profile run?

Comment (yma11) [reply vllm/worker/hpu_enc_dec_model_runner.py:359]: Following cuda path, will it be a problem?

Comment (yma11) [reply vllm/worker/hpu_enc_dec_model_runner.py]: I checked this decorator doesn't exist in `hpu_model_runner`, remove it.

File: vllm/attention/backends/hpu_attn.py [hunk 13]
```diff
@@ -176,11 +176,11 @@ def __init__(
             assert alibi_slopes is None, \
                 'Non-contiguous PA not supported with alibi slopes!'
 
-        suppored_head_sizes = HPUPagedAttention.get_supported_head_sizes()
-        if head_size not in suppored_head_sizes:
+        supported_head_sizes = HPUPagedAttention.get_supported_head_sizes()
```
Comment (michalkuligowski) [inline vllm/attention/backends/hpu_attn.py:179]: Failing tests, blocked on SW-218309
